src/core_codemods/sql_parameterization.py

```diff
@@ -0,0 +1,484 @@
+            cst.CSTNode, cst.CSTNode | cst.RemovalSentinel | cst.FlattenSentinel
+        ] = {}
+        Codemod.__init__(self, context)
+        BaseCodemod.__init__(self, *codemod_args)
+
+    def _build_param_element(self, middle, index: int) -> cst.BaseExpression:
+        # TODO maybe a parameterized string would be better here
+        # f-strings need python 3.6 though
+        if index == 0:
+            return middle[0]
+        operator = cst.Add(
+            whitespace_after=cst.SimpleWhitespace(" "),
+            whitespace_before=cst.SimpleWhitespace(" "),
+        )
+        return cst.BinaryOperation(
+            operator=operator,
+            left=self._build_param_element(middle, index - 1),
+            right=middle[index],
+        )
+
+    def transform_module_impl(self, tree: cst.Module) -> cst.Module:
+        find_queries = FindQueryCalls(self.context)
+        tree.visit(find_queries)
+
+        result = tree
+        for call, query in find_queries.calls.items():
+            ep = ExtractParameters(self.context, query)
+            tree.visit(ep)
+
+            # build tuple elements and fix injection
+            params_elements: list[cst.Element] = []
+            for start, middle, end in ep.injection_patterns:
+                # TODO support for elements from f-strings?
+                # reminder that python has no implicit conversion while concatenating with +, might need to use str() for a particular expression
+                expr = self._build_param_element(middle, len(middle) - 1)
+                params_elements.append(
+                    cst.Element(
+                        value=expr,
+                        comma=cst.Comma(whitespace_after=SimpleWhitespace(" ")),
+                    )
+                )
+                self._fix_injection(start, middle, end)
+
+            # TODO research if named parameters are widely supported
+            # it could solve for the case of existing parameters
+            if params_elements:
+                tuple_arg = cst.Arg(cst.Tuple(elements=params_elements))
+                # self.changed_nodes[call] = call.with_changes(args=[*call.args, tuple_arg])
+                self.changed_nodes[call] = {"args": Append([tuple_arg])}
+
+            # made changes
+            if self.changed_nodes:
+                result = result.visit(ReplaceNodes(self.changed_nodes))
+                self.changed_nodes = {}
+                line_number = self.get_metadata(PositionProvider, call).start.line
+                self.file_context.codemod_changes.append(
+                    Change(
+                        str(line_number), SQLQueryParameterization.CHANGE_DESCRIPTION
+                    ).to_json()
+                )
+                result = result.visit(RemoveEmptyStringConcatenation())
+
+        return result
+
+    # Replace node entirelly -> CSTNode
+    # update some attribute -> dict[str,any]
+    # apend or prepend an argument -> Append([]) / Preprend([])
```

Comment:
move # into docstring

Reply:
These were just some random notes that I was jotting down while thinking on how to solve a problem that originated within the method below. The solution is now the `ReplaceNodes` transform in the same file.

I've deleted those and moved the explanation to `ReplaceNodes` docstring.